src/main/java/com/mongodb/hibernate/jdbc/MongoParameterRecognizer.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2025-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.jdbc;
+
+class MongoParameterRecognizer {
+
+    static String replace(String json) {
+        StringBuilder builder = new StringBuilder(json.length());
+
+        int i = 0;
+        while (i < json.length()) {
+            char c = json.charAt(i++);
+            switch (c) {
+                case '{':
+                case '}':
+                case '[':
+                case ']':
+                case ':':
+                case ',':
+                case ' ':
+                    builder.append(c);
+                    break;
+                case '\'':
+                case '"':
+                    i = scanString(c, i, json, builder);
+                    break;
+                case '?':
+                    builder.append("{$undefined: true}");
```

Comment:
do we need to take into consideration the possibility that `?` shows up in a quoted string? E.g.
```
{
    aggregate: "books",
    pipeline: [
        { $match :  { title: { $eq: "Who's Afraid of Virginia Woolf?" } } },
        { $project: { _id: 1, publishYear: 1, title: 1, author: 1 } }
    ]
}
```

Reply:
This is already account for via the `scanString` method.

Needless to say, we should add comprehensive unit testing for this.